src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java

```diff
@@ -1266,27 +1273,98 @@ private void mergeFirstPartyDataFromUser(ExtUser userExt, ObjectNode result) {
+            segmentIds.forEach(array::add);
+        });
+    }
+
+    private static Map<String, List<String>> groupBySegtaxValues(Map<Integer, List<String>> segments,
+                                                                  Set<Integer> segtaxValues) {
+
+        return segments.entrySet().stream()
+                .map(entry -> Map.entry(resolveSegmentName(entry.getKey(), segtaxValues), entry.getValue()))
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (value1, value2) -> Stream.concat(value1.stream(), value2.stream())
+                                .collect(Collectors.toCollection(ArrayList::new))));
+    }
+
+    private static String resolveSegmentName(Integer taxonomyId, Set<Integer> segtaxValues) {
+        return segtaxValues.contains(taxonomyId) ? SEGTAX_IAB : SEGTAX_TAX + taxonomyId;
+    }
 
-        if (CollectionUtils.isNotEmpty(iabValue)) {
-            final ArrayNode iab = target.putArray("iab");
-            iabValue.forEach(iab::add);
+    private static Map.Entry<Integer, List<Segment>> getValidSegments(Data data) {
+        final ObjectNode ext = data.getExt();
+        final JsonNode taxonomyId = ext != null ? ext.get(SEGTAX) : null;
+
+        if (taxonomyId != null && taxonomyId.isInt()) {
+            final List<Segment> segments = getValidOnlySegments(data.getSegment());
+            if (CollectionUtils.isNotEmpty(segments)) {
+                return Map.entry(taxonomyId.intValue(), segments);
+            }
+        }
+
+        return null;
+    }
+
+    private static List<Segment> getValidOnlySegments(List<Segment> segments) {
+        return segments.stream()
+                .filter(segment -> StringUtils.isNotBlank(segment.getId()))
+                .collect(Collectors.toCollection(ArrayList::new));
+    }
+
+    private static Map<Integer, List<String>> pickRelevantSegments(final Map<Integer, List<Segment>> segments) {
+        final Map<Integer, List<String>> result = new HashMap<>();
+        final List<Integer> segmentsKeys = new ArrayList<>(segments.keySet());
+        segmentsKeys.forEach(key -> result.put(key, new ArrayList<>()));
+
+        for (int i = 0; i < MAX_NUMBER_OF_SEGMENTS; i++) {
+            final int segmentsIndex = i % segmentsKeys.size();
+            final Integer segmentKey = segmentsKeys.get(segmentsIndex);
+            final List<Segment> currentSegments = segments.get(segmentKey);
+
+            result.computeIfPresent(segmentKey, (key, value) -> {
+                value.add(getAndRemoveLastSegment(currentSegments).getId());
+                return value;
+            });
+
+            if (CollectionUtils.isEmpty(currentSegments)) {
+                segmentsKeys.remove(segmentKey);
+                if (CollectionUtils.isEmpty(segmentsKeys)) {
+                    break;
+                }
+            }
```

Comment:
In fact, there is a bug in the current implementation. We want to keep the segments for each key uniformly distributed if possible, but... Let's look at this with an example:

We have the next `keys to segmentsIds` pairs:
```
key1 - value1_1, value1_2, ..., value1_15
key2 - value2_1, ..., value1_14
...
key15 - value15_1
UNLUCKY_KEY - unlucky_value0,  unlucky_value1, ..., (to inf) 
```

So, while we are at first iteration:
- we are picking from key1, key2, ... key14
- now we are picking key15, and deleting it from `segmentsKeys` that causes `i % segmentsKeys.size() == 0` each time we are deleting this pre-last entry
- because of that, we skipped  UNLUCKY_KEY, and any of unlucky_valueX wan't be in picked values